openshift/templates/api/build.yaml

```diff
@@ -0,0 +1,202 @@
+    tags: pmo,api,gdx,agreements-tracker
+parameters:
+  - name: APP_NAME
+    displayName: "App Name"
+    description: "The name of the application (grouped)."
+    required: true
+    value: "gdx-agreements-tracker"
+  - name: ROLE_NAME
+    displayName: "Component Name"
+    description: "The name of the application role."
+    required: true
+    value: "api"
+  - name: PROJECT_NAMESPACE
+    displayName: "OpenShift Project Namespace"
+    description: "The namespace of the OpenShift project containing the application."
+    required: true
+    value: "acd38d"
+  - name: ENV_NAME
+    displayName: "Environment name"
+    description: "The name for this environment [dev, test, prod, tools]"
+    required: true
+    value: "tools"
+
+  - name: BUILDIMAGE_NAME
+    displayName: "Source Image Name"
+    description: "The image to use as the Builder Image."
+    value: "nodejs-16"
+  - name: BUILDIMAGE_TAG
+    displayName: "Source Image Tag"
+    description: "The s2i image tag which is used to build the code."
+    value: "latest"
+
+  - name: RUNTIMEIMAGE_NAME
+    displayName: "Source Image Name"
+    description: "The image to use as the Runtime Image."
+    value: "nodejs-16"
+  - name: RUNTIMEIMAGE_TAG
+    displayName: "Source Image Tag"
+    description: "The s2i image tag which is used to run the code."
+    value: "latest"
+
+  - name: GIT_URL
+    displayName: "GIT Source Repo URL"
+    description: "A GIT URL to your source code."
+    required: true
+    value: "https://github.com/bcgov/gdx-agreements-tracker-api.git"
+  - name: GIT_REF
+    displayName: "Git Reference"
+    description: "The git reference or branch."
+    required: true
+    value: "development"
+  - name: SOURCE_CONTEXT_DIR
+    displayName: "Source Context Directory"
+    description: "The source context directory."
+    required: false
+    value: "api"
+
+  - name: OUTPUT_IMAGE_TAG
+    displayName: "Output Image Tag"
+    description: "The tag given to the built image."
+    required: true
+    value: "dev"
+
+  - name: CPU_LIMIT
+    displayName: "Resources CPU Limit"
+    description: "The resources CPU limit (in cores) for this build."
+    value: "1"
+  - name: MEMORY_LIMIT
+    displayName: "Memory Limit"
+    description: "Maximum amount of memory the container can use."
+    value: "6Gi"
+
+objects:
+  # Image stream for the built base image.
+  - kind: ImageStream
+    apiVersion: v1
+    metadata:
+      name: ${APP_NAME}-${ROLE_NAME}-build
+      namespace: ${PROJECT_NAMESPACE}-${ENV_NAME}
+      annotations:
+        description: Keeps track of changes to the intermediate frontend artifacts image
+      labels:
+        name: ${APP_NAME}-${ROLE_NAME}-build
+        app: ${APP_NAME}
+        role: ${ROLE_NAME}
+        env: ${ENV_NAME}
+
+  # The build config that will be created and be named for the branch you created it for.
+  - kind: BuildConfig
+    apiVersion: v1
+    metadata:
+      name: ${APP_NAME}-${ROLE_NAME}-build.${GIT_REF}
```

Comment:
this is from PIMS and i thought it was a neat idea. if we wanted to do auto-deploys of tagged experimental branches in dev, this would be the first step i think.

Reply:
> 
not sure the extra value of the branch name has here?, doesn't need to be changed at this point.

Reply:
imagine we had 3 tags, one for each of us, experiment1, experiment2, experiment3. say you wanted to demo an old commit (there was different behaviour before, perhaps) without interfering with anything already deployed. we can take these scripts, change the gitref parameter to experiment1, and all of the builds, routes, whathaveyou all get stood up in a few minutes, you tag the old commit with experiment1, and then you can hit experiment1.dev.apps.devops.yaddayadda, and be checking it out in just a few minutes.

this part here is the first piece of the puzzle, and with a handful changes to the scripts we could have that. we can implement that later though, if we want it.

Reply:
Let's hold up a minute. I am concerned that this is getting way to complex figuring out solutions to things that may or may not be an issue in the future.

Let's focus on MVP rather than perfection.

Reply:
to be fair, @SteveMHoward  we are focusing on mvp, this is part of the pr process.  As mentioned in the main comments in the review, the only thing i needed changed was the postgres meta data, I was fine with everything else.

Reply:
i am going to leave as-is then. it already works, doesn't hurt anything, it was free from PIMS, might help us in the future, and is more work to change it now anyway.